src/asl.ts

```diff
@@ -5526,28 +5253,1768 @@ export namespace ASLGraph {
+      usage.type === "Condition" ||
+        usage.type === "StateInput" ||
+        usage.type === "StateInputProps" ||
+        usage.type === "ReturnUsage"
+      ? false
+      : assertNever(usage);
+  }
+
+  function isVariableUsage(variable: string, usage: VariableUsage) {
+    return usage.type === "Assignment" ||
+      usage.type === "Intrinsic" ||
+      usage.type === "StateInput" ||
+      usage.type === "StateInputProps" ||
+      usage.type === "Condition" ||
+      usage.type === "PropAssignment" ||
+      usage.type === "Filter" ||
+      usage.type === "FilterPropAssignment" ||
+      usage.type === "ReturnUsage"
+      ? jsonPathStartsWith(usage.from, variable)
+      : // stateoutput and literal assignments cannot use other variables
+      usage.type === "LiteralAssignment" ||
+        usage.type === "LiteralPropAssignment" ||
+        usage.type === "StateOutput"
+      ? false
+      : assertNever(usage);
+  }
+
+  function analyzeVariables(
+    startState: string,
+    states: States
+  ): {
+    variableUsages: StateVariableUsage[];
+    variableNames: string[];
+    stats: VariableStats[];
+  } {
+    const variableUsages: StateVariableUsage[] = [];
+    const variableNames = new Set<string>();
+    const visited = new Set<string>();
+
+    const topo = topoSort(startState, states);
+
+    depthFirst(startState);
+
+    const variableNamesArray = [...variableNames];
+    // grab all of the variable prefixes.
+    const variableNamePrefixes = [
+      ...new Set(
+        variableNamesArray
+          // ignore the root `$`
+          .filter((v) => v !== "$")
+          .map((v) => normalizeJsonPath(v)[0]!)
+          .map((v) => formatJsonPath([v]))
+      ),
+    ];
+
+    const variableUsageWithId = variableUsages.map((u, i) => ({
+      id: i,
+      index: topo.findIndex((t) => t[0] === u.state),
+      usage: u,
+    }));
+
+    const stats = variableNamePrefixes.map((v) => ({
+      variable: v,
+      assigns: variableUsageWithId.filter(({ usage }) =>
+        isVariableAssignment(v, usage)
+      ),
+      usage: variableUsageWithId.filter(({ usage }) =>
+        isVariableUsage(v, usage)
+      ),
+    }));
+
+    return { variableUsages, variableNames: [...variableNames], stats };
+
+    function depthFirst(stateName: string) {
+      if (visited.has(stateName)) return;
+      visited.add(stateName);
+      const state = states[stateName]!;
+      variableUsages.push(
+        ...getVariableUsage(state).map((usage) => ({
+          ...usage,
+          state: stateName,
+        }))
+      );
+      const names = getVariableNamesFromState(state);
+      names.forEach((n) => variableNames.add(n));
+      visitTransition(state, depthFirst);
+    }
+  }
+
+  function getVariableUsage(state: State): VariableUsage[] {
+    const usages: VariableUsage[] = (() => {
+      if (isPassState(state)) {
+        if (state.ResultPath) {
+          if (state.Result !== undefined) {
```

Comment:
Yep, here we are checking Result is defined. Previously when finding empty Pass states you were checking Result is truthy. Make sure you treat them consistently.

Reply:
In that case it didn't matter because one of them would need to be truthy for the falsey to matter. In this case it matters because the result path is truthy.

Reply:
Finding it hard to follow this reasoning ... but not sure what would be a better approach. Maybe add helpers, `isXYZ` that make the code more expressive?